rlpx/rlpx.go

```diff
@@ -0,0 +1,439 @@
+}
+
+type mstate struct {
+	block  cipher.Block
+	hash   hash.Hash
+	stream cipher.Stream
+}
+
+func newmstate(aesSecret, macSecret []byte) (*mstate, error) {
+	eb, err := aes.NewCipher(aesSecret)
+	if err != nil {
+		return nil, err
+	}
+	es := cipher.NewCTR(eb, make([]byte, 16))
+
+	mb, err := aes.NewCipher(macSecret)
+	if err != nil {
+		return nil, err
+	}
+	h := sha3.NewLegacyKeccak256()
+	return &mstate{
+		hash:   h,
+		block:  mb,
+		stream: es,
+	}, nil
+}
+
+// updates hash using the following devp2p construction:
+//
+// header-mac-seed = aes(mac-secret, keccak256.digest(egress-mac)[:16]) ^ header-ciphertext
+// egress-mac = keccak256.update(egress-mac, header-mac-seed)
+// header-mac = keccak256.digest(egress-mac)[:16]
+func (ms *mstate) header(h []byte) []byte {
+	prev := ms.hash.Sum(nil)
+	dest := make([]byte, 16)
+	ms.block.Encrypt(dest, prev[:16])
+	for i := range dest {
+		dest[i] ^= h[i]
+	}
+	ms.hash.Write(dest)
+	return ms.hash.Sum(nil)[:16]
+}
+
+// updates hash using the following devp2p construction:
+//
+// egress-mac = keccak256.update(egress-mac, frame-ciphertext)
+// frame-mac-seed = aes(mac-secret, keccak256.digest(egress-mac)[:16]) ^ keccak256.digest(egress-mac)[:16]
+// egress-mac = keccak256.update(egress-mac, frame-mac-seed)
+// frame-mac = keccak256.digest(egress-mac)[:16]
+func (ms *mstate) frame(fct []byte) []byte {
+	ms.hash.Write(fct)
+	prev := ms.hash.Sum(nil)
+	dest := make([]byte, 16)
+	ms.block.Encrypt(dest, prev[:16])
+	for i := range dest {
+		dest[i] ^= prev[i]
+	}
+	ms.hash.Write(dest)
+	return ms.hash.Sum(nil)[:16]
+}
+
+func (s *session) decode(buf []byte) (byte, rlp.Item, error) {
+	if len(buf) < 32 {
+		return 0, rlp.Item{}, errors.New("buf too small for 32byte header")
+	}
+	want := s.ig.header(buf[:16])
+	if !hmac.Equal(want, buf[16:32]) {
+		return 0, rlp.Item{}, errors.New("invalid header mac")
+	}
+	s.ig.stream.XORKeyStream(buf[:16], buf[:16])
+	var (
+		frameSize = bint.Decode(buf[:3])
+		frameEnd  = 32 + frameSize + (16 - frameSize%16)
+		frame     = buf[32:frameEnd]
+		frameMac  = buf[frameEnd : frameEnd+16]
+	)
+	if !hmac.Equal(s.ig.frame(frame), frameMac) {
+		return 0, rlp.Item{}, errors.New("invalid frame mac")
+	}
+	s.ig.stream.XORKeyStream(frame, frame)
+	code := frame[0:1]
+	item, err := rlp.Decode(frame[1:frameSize])
+	return code[0], item, isxerrors.Errorf("rlp decoding frame: %w", err)
+}
+
+// Encodees data into an RLPx frame
+//
+// frame = header-ciphertext || header-mac || frame-ciphertext || frame-mac
+// header-ciphertext = aes(aes-secret, header)
+// header = frame-size || header-data || header-padding
+// header-data = [capability-id, context-id]
+// capability-id = integer, always zero
+// context-id = integer, always zero
+// header-padding = zero-fill header to 16-byte boundary
+// frame-ciphertext = aes(aes-secret, frame-data || frame-padding)
+// frame-padding = zero-fill frame-data to 16-byte boundary
+func (s *session) encode(id byte, data []byte) []byte {
```

Comment:
Is the `id` parameter here the message ID (ie the type)? perhaps rename to clarify that?

Reply:
fixed!